snuba/clickhouse/formatter/query.py

```diff
@@ -1,174 +1,251 @@
-    like we did for the AST expression but this is not possible
-    without introducing either circular dependencies or dropping
-    type checking since the visitor and the classes to visit cannot
-    be in the same module as of now and in a visitor pattern they would
-    have to be mutually dependent.
+def _format_composite_query_content(
+    query: CompositeQuery[Table],
+) -> Sequence[FormattedNode]:
+    """
+    Produces the content for composite query (thus no prewhere).
+    """
+    return _format_query_content(
+        [
+            partial(_format_select, query),
+            lambda _: PaddingNode("FROM", format_data_source(query.get_from_clause())),
+            partial(_format_arrayjoin, query),
+            partial(_format_condition, query),
+            partial(_format_groupby, query),
+            partial(_format_having, query),
+            partial(_format_orderby, query),
+            partial(_format_limitby, query),
+            partial(_format_limit, query),
+        ]
+    )
+
+
+def _format_query_content(
+    components: Sequence[
+        Callable[[ClickhouseExpressionFormatter], Optional[FormattedNode]]
+    ]
+) -> Sequence[FormattedNode]:
+    """
+    Takes a sequence of rules to produce the clauses of the query,
+    formats them and returns a Sequence of those that were not None
+    in the right order to compose the query string.
     """
     parsing_context = ParsingContext()
     formatter = ClickhouseExpressionFormatter(parsing_context)
 
+    ret = []
+    for component in components:
+        formatted = component(formatter)
+        if formatted is not None:
+            ret.append(formatted)
+    return ret
+
+
+def _format_select(
+    query: AbstractQuery, formatter: ClickhouseExpressionFormatter
+) -> StringNode:
     selected_cols = [
         e.expression.accept(formatter) for e in query.get_selected_columns_from_ast()
     ]
-    select_clause = StringNode(f"SELECT {', '.join(selected_cols)}")
+    return StringNode(f"SELECT {', '.join(selected_cols)}")
 
-    from_clause = PaddingNode("FROM", format_data_source(query.get_from_clause()))
 
-    ast_arrayjoin = query.get_arrayjoin_from_ast()
-    array_join_clause = (
-        StringNode(f"ARRAY JOIN {ast_arrayjoin.accept(formatter)}")
-        if ast_arrayjoin is not None
+def _build_optional_string_node(
+    name: str,
+    expression: Optional[Expression],
+    formatter: ClickhouseExpressionFormatter,
+) -> Optional[StringNode]:
+    return (
+        StringNode(f"{name} {expression.accept(formatter)}")
+        if expression is not None
         else None
     )
 
-    ast_prewhere = query.get_prewhere_ast()
-    prewhere_clause = (
-        StringNode(f"PREWHERE {ast_prewhere.accept(formatter)}")
-        if ast_prewhere is not None
-        else None
+
+def _format_arrayjoin(
+    query: AbstractQuery, formatter: ClickhouseExpressionFormatter
+) -> Optional[StringNode]:
+    return _build_optional_string_node(
```

Comment:
I don't know if I see the need to wrap all of these parts (arrayjoin, prewhere, condition etc.) in separate functions. You could call `_build_optional_string_node` directly in the `_format_..` functions.

Reply:
I am ok removing this step.